### ISSUE TITLE
apiserver: Request for a custom resource leads to > 1000% CPU usage increase for more than 10min

### ISSUE THREAD

**Issue by ialidzhikov** [kubernetes/kubernetes#109099]
### What happened?

We observe a high CPU usage of apiserver replicas while other replicas are idling:

```
$ k top po -l role=apiserver
kube-apiserver-786785f6fc-q2fwc               49m          737Mi
kube-apiserver-786785f6fc-r2bxz               50m          690Mi
kube-apiserver-786785f6fc-tbph4               11387m       898Mi
kube-apiserver-786785f6fc-zjfvj               35m          544Mi
```

After investigations we managed to drill down to a custom resource create/update requests that cause this high CPU usage (See the steps to reproduce for more details). It makes sense to mention that the above `k top po` output is from a single Node cluster.

We observe that after such request, the CPU usage does not cool down in seconds or a minute. We see that an apiserver replica is affected by high CPU usage for 10-20min. So it takes 10-20min until the apiserver replica's CPU usage goes down again to values from before the request.

Note also the huge imbalance between the apiserver replica's CPU usage. The replica that get's the requests from the client that operates on the custom resources vs the other 3 replicas that are simply idling wrt CPU usage.

### What did you expect to happen?

I would expect a create request for a custom resource to do not fail with a timeout after 30 seconds.
I would expect such request to do not cause > 1000% CPU usage increase for the concrete apiserver replica for more than 10min.

### How can we reproduce it (as minimally and precisely as possible)?

Example CRD and CR can be found in https://gist.github.com/ialidzhikov/a27fdcea78bd3a0fe726b25c0dfc53c0.

1. Create the `clusterserviceplans.servicecatalog.k8s.io` CRD

   ```
   $ k create -f https://gist.githubusercontent.com/ialidzhikov/a27fdcea78bd3a0fe726b25c0dfc53c0/raw/244d78024c31b3ebd870319f83fe53455f2b374c/clusterserviceplans-crd.yaml
   ```
   
2. Try to create a CR:

   ```
   $ k create -f https://gist.githubusercontent.com/ialidzhikov/a27fdcea78bd3a0fe726b25c0dfc53c0/raw/244d78024c31b3ebd870319f83fe53455f2b374c/clusterserviceplan-sample.json --dry-run=server
   ```
   
   Make sure that this create requests times out after >30 seconds with:
   ```
   (on K8s 1.21.10)
   Error from server (Timeout): error when creating "clusterserviceplan-sample.json": Timeout: request did not complete within requested timeout context deadline exceeded
   ```
   ```
   (on K8s 1.20.15)
   Error from server (Timeout): error when creating "clusterserviceplan-sample.json": Timeout: request did not complete within requested timeout 34s
   ```
   
3. Make sure that the apiserver replicas CPU usage is increased by 1000%.

   ![Screenshot 2022-03-29 at 11 38 39](https://user-images.githubusercontent.com/9372594/160570280-2795a61f-9d02-4f28-bc51-22124d6ae231.png)
   
   The graph confirms the CPU spike. Before the dry-run create request the apiserver replicas CPU usage was avg. `40-50m`. After the create request it went up to more than `1200m`. And this high CPU usage lasts for around 10-15min and it takes apiserver to cool down such high amount of time. The apiserver container does not have any cpu limits set.
   Note that this graph is from an idling cluster and the graph demonstrates the reflections on the CPU usage of single apiserver replica for ONLY 1 dry-run create request. Now, image the CPU usage on a busy cluster with a controller that operates on such custom resources.

### Anything else we need to know?

Maybe related issue - https://github.com/kubernetes/kubernetes/issues/102749. But we are already running K8s versions that have the fix for this issue.

I am able to reproduce the issue on 1.18.20, 1.19.16, 1.20.15, 1.21.10, 1.22.7 and 1.23.4.
I am NOT able to reproduce on 1.17.17.

### Kubernetes version

<details>

```console
$ kubectl version

Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.10", GitCommit:"a7a32748b5c60445c4c7ee904caf01b91f2dbb71", GitTreeState:"clean", BuildDate:"2022-02-16T11:18:16Z", GoVersion:"go1.16.14", Compiler:"gc", Platform:"linux/amd64"
```

I am able to reproduce the issue on 1.18.20, 1.19.16, 1.20.15, 1.21.10, 1.22.7 and 1.23.4.
I am NOT able to reproduce on 1.17.17.
</details>

### Cloud provider

<details>
AWS, but I don't think it really matters for the context of this issue.
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>
Gardener
</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by jpbetz** [comment]
Thank you for the detailed report @ialidzhikov. Did you see anything in the apiserver logs that might be related to the CPU utilization?

cc @apelisse

**Comment by jpbetz** [comment]
@apelisse is this maybe another case of the O(2^n) merge issue? There is a largish CR with no schema and with enough levels of depth to maybe trigger that.

**Comment by ialidzhikov** [comment]
> Did you see anything in the apiserver logs that might be related to the CPU utilization?

The logs related to such request seems to be:
```
I0330 10:59:43.959660       1 store.go:1450] "Monitoring resource count at path" resource="clusterserviceplans.servicecatalog.k8s.io" path="<storage-prefix>//servicecatalog.k8s.io/clusterserviceplans"
I0330 10:59:43.960641       1 cacher.go:410] cacher (*unstructured.Unstructured): initialized
I0330 11:00:17.961177       1 trace.go:205] Trace[767094030]: "Create" url:/apis/servicecatalog.k8s.io/v1beta1/clusterserviceplans,user-agent:kubectl/v1.21.0 (darwin/amd64) kubernetes/cb303e6,audit-id:f4be50a3-5815-43e0-9d40-7a8142a68dc0,client:10.243.138.114,accept:application/json,protocol:HTTP/2.0 (30-Mar-2022 10:59:43.960) (total time: 34001ms):
Trace[767094030]: [34.001056475s] [34.001056475s] END
```

After ~5min there is also the following interesting log:
```
E0330 11:05:17.961927       1 finisher.go:157] FinishRequest: post-timeout activity, waited for 5m0.000864197s, child goroutine has not returned yet
```

It probably makes sense to mention that the examples from https://github.com/kubernetes/kubernetes/issues/109099#issue-1184529436 are against apiservers Pods without a CPU limit. We initially noticed the issue on apiserver with a cpu limit of `1` core. In this case the Pod was constantly CPU throttled and a lot of requests were failing sporadically (as expected). When we removed the CPU limit to see how far it goes, (as shared in https://github.com/kubernetes/kubernetes/issues/109099#issue-1184529436) we see that the actual usage goes up to `12` cores.

**Comment by jpbetz** [comment]
Okay, I'll let @apelisse confirm, but this has the hallmarks of an issue discovered recently. I'll try to describe it:

When a CRD uses `preserveUnknownFields: true` in the OpenAPIv3 schema, Server Side Apply refers to the contents as a "deduced type".  There is an issue where when some operations are performed on a deduced type, Server Side Apply performs a compare of the custom resource against itself recursively which is an O(2^n) operation where n is the levels of nesting in custom resource (IIRC the relevant code is https://github.com/kubernetes-sigs/structured-merge-diff/blob/65611e5f7fc60e909529889c29faa9cac17cb485/typed/merge.go#L81-L82).

The CRD in this example has around 20 levels of nesting, which seems like enough to trigger this issue.

The most obvious mitigation that comes to mind to me: Instead of `preserveUnknownFields: true` fully declare the schema of the custom resource using a [structural schema](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/).

**Comment by alexzielenski** [comment]
This is fixed by #111557

I confirmed by following the steps in the original post. I've attach a CPU profile capturing the application of the CR to the apiserver. Also I've attached some flame graphs for those not so inclined:

[before_after_cpuprofiles.zip](https://github.com/kubernetes/kubernetes/files/9224155/before_after_cpuprofiles.zip)

k8s at e6a60c3b4d941da805d7e5548ffd462e5babc74d
<img width="1611" alt="Screen Shot 2022-07-29 at 7 08 24 PM" src="https://user-images.githubusercontent.com/351783/181866346-770d3af2-acee-402b-892f-6d34c32e0f4c.png">

after applying #111557 the `mergingWalker` no longer makes an appearance in any cpu samples
<img width="1610" alt="Screen Shot 2022-07-29 at 7 09 43 PM" src="https://user-images.githubusercontent.com/351783/181866353-2b58f1b1-20ce-486b-8ad4-43ccd27d8fd4.png">
